### PR TITLE
Fix a possible crash in BN_from_montgomery_word

### PR DESCRIPTION
**Pull request by bernd-edlinger**
Thanks to Darovskikh Andrei for for reporting this issue.

Fixes: #5785

##### Checklist

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by bernd-edlinger** [comment]
I wrote the wrong issue in the commit message. will fix that in a moment.

**Comment by dot-asm** [comment]
I [argue against](https://github.com/openssl/openssl/issues/5785#issuecomment-377191261) this.

File: crypto/bn/bn_mont.c [hunk 1]
```diff
@@ -95,8 +95,10 @@ static int BN_from_montgomery_word(BIGNUM *ret, BIGNUM *r, BN_MONT_CTX *mont)
 
     /* clear the top words of T */
     i = max - r->top;
-    if (i)
+    if (i > 0)
         memset(&rp[r->top], 0, sizeof(*rp) * i);
+    else if (i < 0)
```
Comment (richsalz) [inline crypto/bn/bn_mont.c]: If you re-ordered the statements, then you avoid the else.

Comment (bernd-edlinger) [reply crypto/bn/bn_mont.c]: do you mean
```
if (i<0)
   return 0;
if (i)
  memset(....)
```

Comment (richsalz) [reply crypto/bn/bn_mont.c]: Yes, exactly that.  Assuming you do so, +1/approved from me.

Comment (bernd-edlinger) [reply crypto/bn/bn_mont.c]: okay, fine.

**Comment by bernd-edlinger** [comment]
Requested changes have been done.
I will merge tomorrow when no one objects.

**Comment by bernd-edlinger** [comment]
Merged, thanks!

**Comment by dot-asm** [comment]
> I will merge tomorrow when no one objects.

I'd say that this is not exactly appropriate timing to leave such assertion. And that is assuming that it was actually "if nobody *else* objects". As I for one did...

**Comment by bernd-edlinger** [comment]
When I got the necessary approvals I usually wait another 24H
for constructive comments or for an OMC member to raise a -1

**Pull request closed without merging**